pkg/operator/hostetcdendpointcontroller/members.go

```diff
@@ -0,0 +1,136 @@
+package hostetcdendpointcontroller
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/pkg/transport"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog"
+
+	ceoapi "github.com/openshift/cluster-etcd-operator/pkg/operator/api"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+const (
+	etcdCertFile      = "/var/run/secrets/etcd-client/tls.crt"
+	etcdKeyFile       = "/var/run/secrets/etcd-client/tls.key"
+	etcdTrustedCAFile = "/var/run/configmaps/etcd-ca/ca-bundle.crt"
+)
+
+type HealthyEtcdMembersGetter interface {
+	GetHealthyEtcdMembers() ([]string, error)
+}
+
+type healthyEtcdMemberGetter struct {
+	operatorConfigClient v1helpers.OperatorClient
+}
+
+func NewHealthyEtcdMemberGetter(operatorConfigClient v1helpers.OperatorClient) HealthyEtcdMembersGetter {
+	return &healthyEtcdMemberGetter{operatorConfigClient}
+}
+
+func (h *healthyEtcdMemberGetter) GetHealthyEtcdMembers() ([]string, error) {
+	member, err := h.EtcdList("members")
+	if err != nil {
+		return nil, err
+	}
+	hostnames := make([]string, 0)
+	for _, m := range member {
+		hostname := getEtcdName(m.PeerURLS)
+		hostnames = append(hostnames, hostname)
+	}
+	return hostnames, nil
+}
+
+func getEtcdName(peerURLs []string) string {
+	for _, peerURL := range peerURLs {
+		if strings.Contains(peerURL, "etcd-") {
+			return strings.TrimPrefix(strings.Split(peerURLs[0], ".")[0], "https://")
+		}
+	}
+	return ""
+}
+
+func (h *healthyEtcdMemberGetter) EtcdList(bucket string) ([]ceoapi.Member, error) {
+	cli, err := h.getEtcdClient()
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := cli.MemberList(context.Background())
```

Comment:
I don't really understand why the `host-etcd` endpoints are any different than `etcd` endpoints as far as the source of truth is concerned. The entire design of CEO is based on observations why would we break that here? I can see what you mean when you say the controllers rely on these endpoints but the truth is OpenShift today runs with a static set of host-endpoints.

I am not sure what we gain, to be honest I am open to the conversation but I don't think we should need to query etcd api for membership. We know the healthy members and if the list contains an unhealthy member we have a client balancer to fail over to the next.

Reply:
@hexfusion sure, so I started out with comparing the diff from etcd CR `cluster.members` field. But since `etcd-bootstrap` member is in `Unknown` state and it has a hardcoded entry in observation config based on `host-etcd` endpoint, this lead to the two controllers racing or cycling. 

The sequence of events is as follows: 
1. Bootstrap successful
2. Teardown gets kicked in, removes the endpoint.
3. hostendpoint controller sees that `etcd-bootstrap` is actually part of the cluster, based on `clusters.members` field and it re-adds the bootstrap endpoint.
4. undesired state

Now, we can hard code to not add etcd-bootstrap again, but I figured using the etcd API would be better because of the following reasons:
1. We would indirectly have to use the etcd API in managing `etcd` endpoint as well through the liveness probe. So, liveness probe gets triggered through the etcd API and observations are seen in Kube via ready or not ready addresses in `etcd` endpoint.
2. Since `host-etcd` endpoint is not managed by the endpoint controller, it is managed by us, taking this via the etcd API will be consistent with the existing design.

Advantages to me:
1. Makes the `host-etcd` endpoint source of truth for calculating storageURLs
2. The source of truth is a reflection from the etcd API, which is always consistent (as long as kube api is up)
3. If and when we get rid of the DNS, we could potentially use this to populate the IP addresses that can be used as endpoints in storageURLs.

IMO the whole idea of a controller should to keep observed config up to date with latest, hence we should not have an unhealthy member in the list of healthy member (even if the we have load balancer). 

Given the consistency in design, do you have see any issues with etcd API for this observation?

Reply:
And regarding controllers rely on this endpoint: I mean the names of the members are used by kube api server controller and cluster etcd operator to calculate storageURLs and the dns is populated through <hostname>.<cluster-dns>. So IMO, this endpoint should not rely on data populated by cluster-etcd-operator to generate the list.

Reply:
Lets start with the problem:

The problem that I see is that we do not know the actual health/status of the bootstrap node. For this reason, we more or less assume it is `Ready` even though we don't know hence `MemberUnknown` status we report.

A few things here, right now we report Bootstrap as first in the list of endpoints. I think one thing we should consider is pushing bootstrap to the end of the list. If we do this we isolate ourselves from ever using the bootstrap node as a client. A cheap trick but as it will be removed I think its not a bad idea in general.

> Teardown gets kicked in, removes the endpoint.

> hostendpoint controller sees that etcd-bootstrap is actually part of the cluster, based on clusters.members field and it re-adds the bootstrap endpoint.

The problem is that we remove bootstrap but then readd. Could we only ever add bootstrap during install and once it is removed it simply can not be added again regardless of observation in getNewAddressSubset? We will never have another member like bootstrap where kubelet is standalone thus static pod is not observable yet we are implementing this to more or less work around it.

To close I **really** don't want to adjust our design because of a corner case that is removed after installation. We use cluster API in the pod but not in the controllers itself. I would like to continue to boast that our operator and controllers data is based on kube observations only.

Reply:
resolving the conversation after the latest commit